pkg/frontend/querymiddleware/model_extra.go

```diff
@@ -331,32 +336,67 @@ func (vs *vectorSampleStream) UnmarshalJSON(b []byte) error {
 }
 
 func (vs vectorSampleStream) MarshalJSON() ([]byte, error) {
-	if len(vs.Samples) != 1 {
-		return nil, fmt.Errorf("vector sample stream should have exactly one sample, got %d", len(vs.Samples))
+	if (len(vs.Samples) == 1) == (len(vs.Histograms) == 1) { // not XOR
```

Comment:
I think this logic might not be quite right - for example, if we have 0 samples and 2 histograms, we won't return an error.

Reply:
Wait, never mind, ignore me :)